### PR TITLE
[Tune] Add option to not override the working directory

### LINKED ISSUE #29128: [tune] Option to not override working_dir
Currently, Tune always sets the working dir to ~/ray_results/<trial_scoped_dir>. This can be very confusing for users that depend on relative paths.

We should (1) Add a top-level run config option like `should_chdir` to make this very visible, (2) improve docs here, and (3) consider making this the default.

Another issue is that TUNE_ORIG_WORKING_DIR isn't pointing to the runtime_env dir, but the dir on the driver, which isn't the right one in this case. It should point to the runtime_env dir (dir on the worker before the chdir).

### PR DESCRIPTION
**Pull request by justinvyu**
## Why are these changes needed?

See https://github.com/ray-project/ray/issues/29128 for more context on the problem.

This PR does the following:
1. **Fix `TUNE_ORIG_WORKING_DIR`** to pull the correct current working directory within the worker process when the Tune trial logdir is being created. This PR deprecates this environment variable as it's confusing to get Tune metadata from sources other than `session`.
2. **Introduce a `chdir_to_trial_dir` flag** in the TuneConfig that defaults to `True`, which configures whether or not Tune should change the working directory of each worker to its corresponding trial directory.
    - If this flag is set to False, the user may still want to access the Tune trial directory. This can be done with a **newly added `session.get_trial_dir()` API.**
3. Make the `TUNE_ORIG_WORKING_DIR` deprecation, `chdir_to_trial_dir` flag, and `session.get_trial_dir()` more visible in the documentation with an **example in the Tune FAQ.**

## Related issue number

Context https://github.com/ray-project/ray/issues/9571
Closes https://github.com/ray-project/ray/issues/29128

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: doc/source/tune/faq.rst [hunk 1]
```diff
@@ -733,3 +733,59 @@ If `grid_search` is provided as an argument, the grid will be repeated ``num_sam
...
+we recommend you to **only perform read operations**. Writes should be performed
+in the independent trial directories. Here are two ways of accomplishing this:
+
+
+Option 1: Setting `chdir_to_log_dir=False` in `air.RunConfig`
+'''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+The first option is to explicitly tell Tune to not change the working directory
+to the trial directory. In this case, the :meth:`session.get_log_dir() <ray.air.session.get_log_dir>`
+API should be used to get the path for writes.
```
Comment (ericl) [inline doc/source/tune/faq.rst:765]:
```suggestion
API should be used to get the path for saving trial-specific outputs.
```

File: doc/source/tune/faq.rst [hunk 2]
```diff
@@ -733,3 +733,59 @@ If `grid_search` is provided as an argument, the grid will be repeated ``num_sam
...
+There are two main reasons for this default:
+
+1. If trial writes artifacts, we want to avoid conflicts: if two trials on the same machine write to ``artifact.bin``, the second one would overwrite the artifact from the first.
+
+2. In distributed settings with more than one machine, we cannot guarantee that the working directory from the driver script exists on the other remote nodes. However, we can guarantee that the trial directory exists.
+
+If you need to access files from the original working directory
+(ex: you need to load a (relatively small) dataset or
+saved models from ``~/code/data/data.csv`` or ``~/code/models/model.h5``),
+we recommend you to **only perform read operations**. Writes should be performed
```
Comment (ericl) [inline doc/source/tune/faq.rst:756]: This paragraph is confusing, I would recommend removing/rewording.

File: doc/source/tune/faq.rst [hunk 3]
```diff
@@ -733,3 +733,59 @@ If `grid_search` is provided as an argument, the grid will be repeated ``num_sam
...
+    :emphasize-lines: 9, 15
+    :language: python
+    :start-after: __no_chdir_start__
+    :end-before: __no_chdir_end__
+
+Option 2: Manual control using environment variables
+''''''''''''''''''''''''''''''''''''''''''''''''''''
+
+The second option is to allow Tune to change the working directory and use the
+``RAY_ORIG_WORKING_DIR`` environment variable to convert relative paths to the
```
Comment (ericl) [inline doc/source/tune/faq.rst:778]: Why don't we add this to the `session` object and deprecate the env vars altogether? It seems inconsistent to have some metadata in env vars and some in the session object.

File: doc/source/tune/faq.rst [hunk 4]
```diff
@@ -733,3 +733,59 @@ If `grid_search` is provided as an argument, the grid will be repeated ``num_sam
...
+.. literalinclude:: doc_code/faq.py
+    :dedent:
+    :emphasize-lines: 2, 11
+    :language: python
+    :start-after: __abspath_with_env_var_start__
+    :end-before: __abspath_with_env_var_end__
+
+It's best not to rely on relative paths regardless, so consider converting
+to absolute paths for this option.
+Example: ``open(Path(session.get_log_dir()) / "write.txt", "w")``
```
Comment (ericl) [inline doc/source/tune/faq.rst:791]: This paragraph didn't make sense to me, consider removing.

File: doc/source/tune/faq.rst [hunk 5]
```diff
@@ -733,3 +733,59 @@ If `grid_search` is provided as an argument, the grid will be repeated ``num_sam
...
+Let's say you launch a Tune experiment from ``~/code/my_script.py``. By default, Tune
+changes the working directory of each worker from ``~/code`` to its corresponding trial
+directory (e.g. ``~/ray_results/exp_name/trial_0000x``). This guarantees separate working
+directories for each worker process.
+
+There are two main reasons for this default:
+
+1. If trial writes artifacts, we want to avoid conflicts: if two trials on the same machine write to ``artifact.bin``, the second one would overwrite the artifact from the first.
+
+2. In distributed settings with more than one machine, we cannot guarantee that the working directory from the driver script exists on the other remote nodes. However, we can guarantee that the trial directory exists.
```
Comment (ericl) [inline doc/source/tune/faq.rst:751]: How about removing these two points? I don't think the level of details is useful for users. The statement that we guarantee separate working directories [to ensure output files don't conflict] is sufficient.

File: python/ray/air/_internal/session.py [hunk 6]
```diff
@@ -74,6 +74,13 @@ def trial_resources(self) -> "PlacementGroupFactory":
         """Trial resources for the corresponding trial."""
         raise NotImplementedError
 
+    @property
+    def log_dir(self) -> str:
```
Comment (ericl) [inline python/ray/air/_internal/session.py:78]: Should this be `trial_dir`?

File: python/ray/air/_internal/session.py [hunk 7]
```diff
@@ -74,6 +74,13 @@ def trial_resources(self) -> "PlacementGroupFactory":
...
+        """Log directory corresponding to the session where data such as checkpoints
+        can be written.
+        """
+        raise NotImplementedError
+
```
Comment (ericl) [inline python/ray/air/_internal/session.py:83]: Add `orig_working_dir`?

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/tune/trainable/trainable.py [hunk 8]
```diff
@@ -305,6 +307,41 @@ def train_buffered(self, buffer_time_s: float, max_buffer_length: int = 1000):
 
         return results
 
+    def _log_custom_error_if_possible(self, exception: Exception):
```
Comment (ericl) [inline python/ray/tune/trainable/trainable.py:310]: Can we split this into a separate PR? This change is quite large already.

File: python/ray/tune/tune.py [hunk 9]
```diff
@@ -525,8 +535,22 @@ def _handle_string_queue():
             "well as implementing `reset_config` for Trainable."
         )
 
+    if not chdir_to_log_dir:
+        logger.warning(
```
Comment (ericl) [inline python/ray/tune/tune.py:539]: This warning is gratuitious; remove.

File: python/ray/air/_internal/session.py (hunk 6 above)
Comment (justinvyu) [reply python/ray/air/_internal/session.py:78]: I believe `log_dir` is more accurate here, since a Train session would return the worker-level log directory within the Tune trial directory. A Tune session would just return the trial dir.

File: doc/source/tune/faq.rst (hunk 3 above)
Comment (justinvyu) [reply doc/source/tune/faq.rst:778]: I was considering this but didn't know whether this 2nd option was too niche of a use case to introduce a high level `session` API for it.

Comment (justinvyu) [reply doc/source/tune/faq.rst:766]: Should we deprecate the environment variable and only recommend option 1 of setting the chdir flag to False? I'm just worried about cluttering the `session` API with `get_orig_working_dir` that might not be as needed as something like `get_log_dir`. cc @krfricke

Comment (ericl) [reply doc/source/tune/faq.rst:766]: Yes, I think we should do this. There shouldn't be multiple ways of getting session metadata. Also, this is much more discoverable than an env var.

File: python/ray/air/_internal/session.py (hunk 6 above)
Comment (ericl) [reply python/ray/air/_internal/session.py:78]: Tune and Train sessions shouldn't have diverging behaviors. How about we call this `trial_dir`, which always returns the trial dir, and then Train can have a separate call for worker-level log dirs?

File: python/ray/tune/impl/tuner_internal.py [hunk 10]
```diff
@@ -77,10 +81,28 @@ def __init__(
...
+                )
+
+            if run_config:
+                # If a run config is passed in, overwrite the Trainer's run config
+                trainable.run_config = run_config
+                trainable._param_dict["run_config"] = run_config
+            else:
+                # If no run config was passed to Tuner directly, use the one from the
+                # Trainer if available
+                run_config = trainable.run_config
```
Comment (ericl) [inline python/ray/tune/impl/tuner_internal.py:105]: This fix seems unrelated, could we put it into a separate PR if so?

File: python/ray/tune/impl/tuner_internal.py [hunk 11]
```diff
@@ -382,6 +404,7 @@ def _get_tune_run_arguments(self, trainable) -> Dict[str, Any]:
             reuse_actors=self._tune_config.reuse_actors,
             max_concurrent_trials=self._tune_config.max_concurrent_trials,
             time_budget_s=self._tune_config.time_budget_s,
+            chdir_to_log_dir=self._run_config.chdir_to_log_dir,
```
Comment (ericl) [inline python/ray/tune/impl/tuner_internal.py:407]:
```suggestion
            chdir_to_trial_dir=self._run_config.chdir_to_trial_dir,
```

File: python/ray/air/_internal/session.py (hunk 6 above)
Comment (justinvyu) [reply python/ray/air/_internal/session.py:78]: For the unified AIR session, should it be `session.get_trial_dir()` and `session.get_log_dir()`? For Tune sessions though, these two would return the same thing.

Comment (ericl) [reply python/ray/air/_internal/session.py:78]: Why don't we just not expose the log dir for now and only trial_dir? This is a new thing right? Train can always ignore the chdir flag anyways.

Actually, I don't think this PR should be having any considerations / impact for Train.

Comment (justinvyu) [reply python/ray/air/_internal/session.py:78]: Train currently also does a `chdir` to the worker-level log dir for checkpointing without contention between workers. I updated the PR to allow configuring Train to not do this chdir (basically having consistent functionality between Train and Tune).

Comment (ericl) [reply python/ray/air/_internal/session.py:78]: How about we implement that in a follow-up then? This PR can add only the Tune changes and only add an accessory for the tune trial dir. We can then add Train support.

File: python/ray/train/data_parallel_trainer.py [hunk 12]
```diff
@@ -338,7 +338,7 @@ def training_loop(self) -> None:
             name=session.get_trial_name(),
             id=session.get_trial_id(),
             resources=session.get_trial_resources(),
-            logdir=os.getcwd(),
+            logdir=session.get_trial_dir(),
```
Comment (justinvyu) [inline python/ray/train/data_parallel_trainer.py:341]: This is still a needed change in this PR, since setting `chdir_to_trial_dir=False` in the Tuner would make `os.getcwd()` return the original working directory here, which is not the trial logdir.

**Review by ericl — APPROVED** [review]
LGTM!

File: python/ray/tune/tune_config.py [hunk 13]
```diff
@@ -59,3 +66,4 @@ class TuneConfig:
     max_concurrent_trials: Optional[int] = None
     time_budget_s: Optional[Union[int, float, datetime.timedelta]] = None
     reuse_actors: Optional[bool] = None
+    chdir_to_trial_dir: bool = True
```
Comment (justinvyu) [inline python/ray/tune/tune_config.py:69]: When I do a follow-up PR to allow Trainers to also not change the working dir, this config would need to be moved to `air.RunConfig` (something shared between Trainers and Tuners).

**Review by krfricke — APPROVED** [review]
LGTM

**Comment by krfricke** [comment]
Re-running the failing GPU test

**Comment by justinvyu** [comment]
> Re-running the failing GPU test

Looks like it's passing now with latest master merged in.
